shell/platform/tizen/tizen_vsync_waiter.cc

```diff
@@ -4,40 +4,103 @@
-  TizenVsyncWaiter* tizen_vsync_waiter =
-      reinterpret_cast<TizenVsyncWaiter*>(data);
-  if (!ecore_thread_check(thread)) {
-    tdm_error error = tdm_client_vblank_wait(tizen_vsync_waiter->vblank_, 1,
-                                             TdmClientVblankCallback, data);
-    if (error != TDM_ERROR_NONE) {
-      FT_LOGE("tdm_client_vblank_wait error  %d", error);
-      return;
-    }
-    tdm_client_handle_events(tizen_vsync_waiter->client_);
-  }
-}
+static const int QUEUE_QUIT = -1;
+static const int QUEUE_REQUEST_VBLNAK = 0;
+
+typedef struct {
+  Eina_Thread_Queue_Msg head;
+  int value;
+} Msg;
+
+static Eina_Thread_Queue* vblankThreadQueue{nullptr};
 
 TizenVsyncWaiter::TizenVsyncWaiter(TizenEmbedderEngine* engine)
     : engine_(engine) {
   if (!CreateTDMVblank()) {
     FT_LOGE("Failed to create TDM vblank");
     DestoryTDMVblank();
+  } else {
+    vblankThreadQueue = eina_thread_queue_new();
+    vblankThread_ =
+        ecore_thread_feedback_run(RequestVblankLoop, NULL, VblankLoopFinish,
+                                  VblankLoopFinish, this, EINA_TRUE);
   }
 }
 
-TizenVsyncWaiter::~TizenVsyncWaiter() { DestoryTDMVblank(); }
+TizenVsyncWaiter::~TizenVsyncWaiter() {
+  SendMessage(QUEUE_QUIT);
+  if (vblankThread_) {
+    ecore_thread_cancel(vblankThread_);
+    vblankThread_ = nullptr;
+  }
+  DestoryTDMVblank();
+}
 
 void TizenVsyncWaiter::AsyncWaitForVsync(intptr_t baton) {
   baton_ = baton;
   if (TDMValid()) {
-    ecore_thread_run(RequestVblank, NULL, NULL, this);
+    SendMessage(QUEUE_REQUEST_VBLNAK);
+  }
+}
+
+void TizenVsyncWaiter::SendMessage(int val) {
+  if (!vblankThreadQueue || !vblankThread_) {
+    FT_LOGE("vblank thread or vblank thread queue not valid");
+    return;
+  }
+  Msg* msg;
+  void* ref;
+  msg = (Msg*)eina_thread_queue_send(vblankThreadQueue, sizeof(Msg), &ref);
+  msg->value = val;
+  eina_thread_queue_send_done(vblankThreadQueue, ref);
+}
+
+void TizenVsyncWaiter::RequestVblankLoop(void* data, Ecore_Thread* thread) {
+  TizenVsyncWaiter* tizen_vsync_waiter =
+      reinterpret_cast<TizenVsyncWaiter*>(data);
+  void* ref;
+  Msg* msg;
+  while (!ecore_thread_check(thread)) {
+    if (!vblankThreadQueue) {
+      FT_LOGE("Vblank thread queue is not valid");
+      return;
+    }
+    msg = (Msg*)eina_thread_queue_wait(vblankThreadQueue, &ref);
+    if (msg) {
+      eina_thread_queue_wait_done(vblankThreadQueue, ref);
+    } else {
+      FT_LOGE("Message is null");
+      continue;
+    }
+    if (msg->value == QUEUE_QUIT) {
+      FT_LOGD("Message queue quit");
+      return;
+    }
+    if (!tizen_vsync_waiter->TDMValid()) {
+      FT_LOGE("TDM not Valid");
+      return;
+    }
+    tdm_error error = tdm_client_vblank_wait(tizen_vsync_waiter->vblank_, 1,
+                                             TdmClientVblankCallback, data);
+    if (error != TDM_ERROR_NONE) {
+      FT_LOGE("tdm_client_vblank_wait error  %d", error);
+      tizen_vsync_waiter->DestoryTDMVblank();
```

Comment:
Is it appropriate to destroy all on any tdm_client_vblank_wait error?

Reply:
I think try to request vblank again, maybe is a better way.

Reply:
Previously you just returned on error without doing anything further. Could that make any problem?

Reply:
If just return, don't destory TDM, we could request vblank again.
if destory TDM before return, we could not request vblank.

Reply:
I mean, in the current code we don't destroy TDM or request vblank again, therefore do nothing on error. Do you mean that's not enough?

Reply:
I think destory TDM is not a good way when meet wait error.
If we don't destory TDM, we have chance render again, so we could request vblank again or do nothing.
=================request vblank again==================
  if (error != TDM_ERROR_NONE) {
      FT_LOGE("tdm_client_vblank_wait error  %d", error);
      tizen_vsync_waiter->SendMessage(QUEUE_REQUEST_VBLANK);
      continue;
    }
=================do nothing==================
  if (error != TDM_ERROR_NONE) {
      FT_LOGE("tdm_client_vblank_wait error  %d", error);
      continue;
    }
==========================================

Reply:
Requesting again without counting number of retries may result in an infinite loop or an overflow of the queue so doing nothing looks better I think.

Reply:
Yes, ok